src/interfaces/cloud.rs

```diff
@@ -0,0 +1,410 @@
+    insert_chains_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+}
+
+impl Display for Token {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut keys = self.findex_master_key.to_vec();
+
+        if let Some(fetch_entries_key) = &self.fetch_entries_key {
+            keys.push(0);
+            keys.extend(fetch_entries_key.as_ref());
+        }
+
+        if let Some(fetch_chains_key) = &self.fetch_chains_key {
+            keys.push(1);
+            keys.extend(fetch_chains_key.as_ref());
+        }
+
+        if let Some(upsert_entries_key) = &self.upsert_entries_key {
+            keys.push(2);
+            keys.extend(upsert_entries_key.as_ref());
+        }
+
+        if let Some(insert_chains_key) = &self.insert_chains_key {
+            keys.push(3);
+            keys.extend(insert_chains_key.as_ref());
+        }
+
+        write!(f, "{}{}", self.index_id, base64::encode(keys))
+    }
+}
+
+impl FromStr for Token {
+    type Err = FindexErr;
+
+    fn from_str(token: &str) -> Result<Self, Self::Err> {
+        let (index_id, tail) = token.split_at(INDEX_ID_LENGTH);
+        let mut bytes = base64::decode(tail)
+            .map_err(|_| {
+                FindexErr::Other(format!(
+                    "token {token} is not a valid base64 encoded string"
+                ))
+            })?
+            .into_iter();
+
+        let findex_master_key = KeyingMaterial::try_from_bytes(
+            &bytes.next_chunk::<MASTER_KEY_LENGTH>().map_err(|_| {
+                FindexErr::Other(
+                    "the token is too short, cannot read the Findex master key".to_owned(),
+                )
+            })?,
+        )?;
+
+        let mut token = Token {
+            index_id: index_id.to_owned(),
+            findex_master_key,
+
+            fetch_entries_key: None,
+            fetch_chains_key: None,
+            upsert_entries_key: None,
+            insert_chains_key: None,
+        };
+
+        while let Some(prefix) = bytes.next() {
+            let key = Some(
+                bytes
+                    .next_chunk::<SIGNATURE_KEY_LENGTH>()
+                    .map_err(|_| {
+                        FindexErr::Other(format!(
+                            "the token is too short, expecting {SIGNATURE_KEY_LENGTH} bytes after \
+                             the prefix {prefix}"
+                        ))
+                    })?
+                    .into(),
+            );
+
+            if prefix == 0 {
+                token.fetch_entries_key = key;
+            } else if prefix == 1 {
+                token.fetch_chains_key = key;
+            } else if prefix == 2 {
+                token.upsert_entries_key = key;
+            } else if prefix == 3 {
+                token.insert_chains_key = key;
+            } else {
+                return Err(FindexErr::Other(format!(
+                    "the token contains a unknown prefix {prefix}"
```

Comment:
Same here, maybe you can give the offset.

Reply:
Done in 424a022